rfcs/proposed/task_group_dynamic_dependencies/implementation_details.md

```diff
@@ -0,0 +1,516 @@
+
+<img src="assets/impl_sequence_diagram_transfer.png">
+
+### Dynamic state lifetime issue
+
+During the implementation of `transfer_this_task_completion_to`, a lifetime management issue was identified in `task_dynamic_state`.
+
+Initially, its reference counter was incremented only when the task or a `task_completion_handle` was created.
+
+Consider the following example: task `A` and a `task_completion_handle` named `A_comp_handle`, initially assigned to handle the completion of `A`. 
+The dynamic state `A_state` is associated with task `A`. 
+
+While task `A` is executing, it transfers its completion to task `B` (illustrated by the red arrow in the diagram below).
+The dynamic state `B_state` is associated with task `B`.
+
+As described earlier, `A_state::m_new_dynamic_state` is set to point to `B_state` (as shown by the green arrow in the diagram).
+
+<img src="assets/impl_lifetime_issue.png" width=800>
+
+After task `A` completes, task `B` is executed and subsequently destroyed, along with its associated `B_state`.
+
+If a new successor is added using `A_comp_handle` after this point, the request will be redirected to `B_state`.
+
+However, since `B_state` was destroyed after task `B` completed, accessing it would lead to use-after-free behavior.
+
+The solution is to extend the lifetime of `B_state` by incrementing its reference counter in every dynamic state that stores it in `m_new_dynamic_state`.
+The counter is then decreased when `A_state` is destroyed.
+
+## Library ABI changes
+
+As previously mentioned, implementing `transfer_this_task_completion_to` requires introducing a new entry point to the TBB binary:
+
+```cpp
+namespace tbb::detail::rN {
+    d1::task* current_task_ptr();
+}
+```
+
+The function returns a pointer to the innermost task currently being executed. If no task is executing, it returns `nullptr`.
+
+When the scheduler retrieves a task for execution from any source, it stores a pointer to that task in the corresponding
+`task_dispatcher` object. This happens before invoking either `task::execute` or `task::cancel`.
+
+When `current_task_ptr()` is called, the pointer to the currently executing task is retrieved from the task dispatcher.
+
+If a different task is retrieved during the execution, for example, when a blocking TBB call is made from within the task body,
+the stored pointer is temporarily replaced and restored once the blocking call completes.
+
+## Open questions to the implementation
+
+The following questions were raised during the review of this this document and should be addressed before promoting the feature to `supported`:
+* When transferring task completion, the end of the successor list is determined by traversing it, which may become a
+  performance concern if the list contains many elements. Storing a tail pointer in `task_dynamic_state` may resolve this, assuming a
+  real-world case confirms the performance concern.
```

Comment:
I am not sure about us having many real-world cases before productization. I would change to "assuming a realistic use case example confirms the performance concerns".